src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+        Span
+        before(Seq spot)
+        {
+            return sub(start_, spot);
+        }
+
+        bool
+        empty() const
+        {
+            return start_ == end_;
+        }
+
+        //Return the ID of the ledger that starts this span
+        ID
+        startID() const
+        {
+            return ledger_[start_];
+        }
+
+        // Return the ledger sequence number of the first possible difference
+        // between this span and a given ledger.
+        Seq
+        diff(Ledger const& o) const
+        {
+            return clamp(mismatch(ledger_, o));
+        }
+
+        //  The Seq and ID of the end of the span
+        std::pair<Seq, ID>
+        tip() const
+        {
+            Seq tipSeq{end_ -Seq{1}};
+            return {tipSeq, ledger_[tipSeq]};
+        }
+
+    private:
+        Span(Seq start, Seq end, Ledger const& l)
+            : start_{start}, end_{end}, ledger_{l}
+        {
+            assert(start <= end);
+        }
+
+        Seq
+        clamp(Seq val) const
+        {
+            return std::min(std::max(start_, val), end_);
+        };
+
+        // Return a span of this over the half-open interval [from,to)
+        Span
+        sub(Seq from, Seq to)
+        {
+            return Span(clamp(from), clamp(to), ledger_);
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream& o, Span const& s)
+        {
+            return o << s.tip().second << "(" << s.start_ << "," << s.end_
+                     << "]";
+        }
+
+        friend Span
+        merge(Span const& a, Span const& b)
+        {
+            // Return combined span, using ledger_ from higher sequence span
+            if (a.end_ < b.end_)
+                return Span(std::min(a.start_, b.start_), b.end_, b.ledger_);
+
+            return Span(std::min(a.start_, b.start_), a.end_, a.ledger_);
+        }
+    };
+
+    // A node in the trie
+    struct Node
+    {
+        Node() : span{}, tipSupport{0}, branchSupport{0}
+        {
+        }
+
+        Node(Ledger const& l) : span{l}, tipSupport{1}, branchSupport{1}
+        {
+        }
+
+        Node(Span s) : span{std::move(s)}
+        {
+        }
+
+        Span span;
+        std::uint32_t tipSupport = 0;
+        std::uint32_t branchSupport = 0;
+
+        std::vector<std::unique_ptr<Node>> children;
+        Node* parent = nullptr;
+
+        /** Remove the given node from this Node's children
+
+            @param child The address of the child node to remove
+            @note The child must be a member of the vector. The passed pointer
+                  will be dangling as a result of this call
+        */
+        void
+        erase(Node const* child)
+        {
+            auto it = std::remove_if(
+                    children.begin(),
+                    children.end(),
+                    [child](std::unique_ptr<Node> const& curr) {
+                        return curr.get() == child;
+                    });
```

Comment:
Use `find` instead of `remove_if` (we know the found element is unique).
Since we don't care about order, we can swap with the end element and `pop_back` as well, instead of calling `erase`.

Reply:
I was thinking the `remove_if` gave us a better check on the logic :  That `child` appears exactly once.  Otoh, we don't need debug checks in release code.  So what about using `find_if`, but following up in an `assert` that `child` doesn't appear after `it`:
```c++
auto it = std::find_if(children.begin(), children.end(),
                       [child](std::unique_ptr<Node> const& curr)
                       {
                        return curr.get() == child;
                       });
assert(it != children.end());
assert(std::none_of(it, children.end(),
                    [child](std::unique_ptr<Node> const& curr)
                    {
                        return curr.get() == child;
                    });
children.erase(it);
```
I stuck with `erase`, instead of `swap`/`pop_back`.  `vector` knows best how to erase a single element out of the middle.

Reply:
@HowardHinnant the contents of the vector being `unique_ptr` is a good guarantee that the children are unique, do we really need another check?

Re: `vector` knows best how to erase a single element out of the middle. Agreed! However, it also needs to keep the elements in relative order. We don't need to do that. I'd expect swapping and `pop_back` to be better; However, I also doubt it'll make that big a difference. I'm fine with `erase` if you prefer it.

Reply:
On the uniqueness of the pointer, just call me paranoid. :-)  I don't have strong feelings about that check.  I'm happy to leave it up to Brad.

Agreed, it won't make a lot of difference on how we erase the single item.  I just figured one call to do it instead of two would make our code a little simpler.  And `vector::erase` won't be more expensive than `swap/pop_back`, and _might_ be just a little faster.  Not a big deal either way.

Reply:
Scott and I talked offline.  If `[it, end())` could grow large, his suggestion of `swap/pop_back` is the way to go.  The definition of "large" is pretty grey, and I'm not sure what it is.  Under 100 and I wouldn't sweat it too much (not worth our time).  Over 1,000,000, and `swap/pop_back` is the clear winner, assuming of course we are free to reorder `children` under this operation.

Reply:
Thanks for the feedback. This will be tracking one validation per configured validator, so in the worst case if all `N`  validators disagree on the current ledger, the vector would have `N` entries. I think that puts us squarely in the "not-large" setting for for the foreseeable future. I'll take both options into account when addressing these comments.